src/layer.rs

```diff
@@ -1022,6 +1022,20 @@ where
             let mut otel_data = span.extensions_mut().remove::<OtelData>();
             let span_builder = otel_data.as_mut().map(|data| &mut data.builder);
 
+            let limits = self.tracer.span_limits();
```

Comment:
could be worth looking at the benchmarks to see what the perf impact of this is. Could alternatively load this when the layer is initialized or otherwise lazy load this value once?

Reply:
Sure, I'll try to run the benchmarks and get back to you hopefully tomorrow.

> Could alternatively load this when the layer is initialized or otherwise lazy load this value once?

Right now, probably. I'm not sure in the long-term though, because the limits may be changed through a `TracerProvider` and all `Tracer`s should honor that change even if they were created earlier. Currently, the limits cannot be changed after a provider is created in opentelemetry-sdk, but they may add it later. I do not know of any other Tracer implementation, but that could also be a consideration.

Reply:
I've added a benchmark with events, see this PR for details.

### Current `HEAD`

```
otel_many_events/spec_baseline
                        time:   [19.951 µs 19.969 µs 19.991 µs]
                        change: [-0.2111% +0.0843% +0.3784%] (p = 0.58 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) high mild
  8 (8.00%) high severe
otel_many_events/no_data_baseline
                        time:   [48.757 µs 48.803 µs 48.859 µs]
                        change: [+0.1993% +0.4747% +0.7119%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) high mild
  10 (10.00%) high severe
otel_many_events/data_only_baseline
                        time:   [68.263 µs 68.286 µs 68.313 µs]
                        change: [+0.2342% +0.5605% +0.8355%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe
otel_many_events/full_filtered
                        time:   [251.47 µs 251.64 µs 251.83 µs]
                        change: [+82.121% +82.453% +82.717%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) high mild
  8 (8.00%) high severe
otel_many_events/full_not_filtered
                        time:   [280.82 µs 280.92 µs 281.02 µs]
                        change: [+2.0043% +2.1354% +2.2589%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 15 outliers among 100 measurements (15.00%)
  2 (2.00%) low severe
  3 (3.00%) low mild
  6 (6.00%) high mild
  4 (4.00%) high severe
```

### This PR

```
otel_many_events/spec_baseline
                        time:   [19.504 µs 19.520 µs 19.540 µs]
                        change: [-2.4723% -2.1766% -1.8365%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 18 outliers among 100 measurements (18.00%)
  8 (8.00%) high mild
  10 (10.00%) high severe
otel_many_events/no_data_baseline
                        time:   [48.471 µs 48.558 µs 48.669 µs]
                        change: [-0.7518% -0.6127% -0.4556%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 16 outliers among 100 measurements (16.00%)
  3 (3.00%) high mild
  13 (13.00%) high severe
otel_many_events/data_only_baseline
                        time:   [67.843 µs 67.891 µs 67.947 µs]
                        change: [-0.6862% -0.5205% -0.3061%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  4 (4.00%) high severe
otel_many_events/full_filtered
                        time:   [136.64 µs 136.72 µs 136.81 µs]
                        change: [-45.679% -45.580% -45.464%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) high mild
  11 (11.00%) high severe
otel_many_events/full_not_filtered
                        time:   [292.62 µs 292.72 µs 292.82 µs]
                        change: [+4.1160% +4.2252% +4.3375%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  3 (3.00%) high mild
  2 (2.00%) high severe

```

The benchmarks do not seem to be very consistent though. It might be just due to background workload on my machine, but the first benchamarks which do not depend on the changes at all have wildly different results so it would be great if someone else would also run it in some better environment.

The default for filtering is 128 events and the benchmark generates 1000 so the filtered result has much better performance because it doesn't need to process the events that would be discarded anyway.

The not filtered benchmark sets the limit to 1000 so nothing is filtered, there is a 4% penalty.